File: assignment 10/python/programs/10.py
from collections import defaultdict


class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def dfs(self, start_node, visited=None):
        if visited is None:
            visited = set()
        visited.add(start_node)
        print(start_node, end=" ")

        for neighbor in self.graph[start_node]:
            if neighbor not in visited:
                self.dfs(neighbor, visited)


# Example usage:
if __name__ == "__main__":
    g = Graph()
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 2)
    g.add_edge(2, 0)
    g.add_edge(2, 3)
    g.add_edge(3, 3)

    start_node = 2  # Start DFS from node 2
    print("DFS Traversal (starting from node 2):")
    g.dfs(start_node)

File: assignment 10/c/programs/10.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_VERTICES 100

// Structure for a graph node
struct Node
{
    int data;
    struct Node *next;
};

// Structure for a graph
struct Graph
{
    int numVertices;
    struct Node **adjacencyList;
    bool *visited;
};

// Function to create a new graph node
struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to create a graph with a given number of vertices
struct Graph *createGraph(int numVertices)
{
    struct Graph *graph = (struct Graph *)malloc(sizeof(struct Graph));
    graph->numVertices = numVertices;
    graph->adjacencyList = (struct Node **)malloc(numVertices * sizeof(struct Node *));
    graph->visited = (bool *)malloc(numVertices * sizeof(bool));

    for (int i = 0; i < numVertices; i++)
    {
        graph->adjacencyList[i] = NULL;
        graph->visited[i] = false;
    }

    return graph;
}

// Function to add an edge to the graph
void addEdge(struct Graph *graph, int src, int dest)
{
    struct Node *newNode = createNode(dest);
    newNode->next = graph->adjacencyList[src];
    graph->adjacencyList[src] = newNode;
}

// Recursive function for DFS traversal
void DFS(struct Graph *graph, int vertex)
{
    graph->visited[vertex] = true;
    printf("%d ", vertex);

    struct Node *temp = graph->adjacencyList[vertex];
    while (temp != NULL)
    {
        int neighbor = temp->data;
        if (!graph->visited[neighbor])
        {
            DFS(graph, neighbor);
        }
        temp = temp->next;
    }
}

int main()
{
    struct Graph *graph = createGraph(7);

    // Add edges to the graph
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 2);
    addEdge(graph, 1, 3);
    addEdge(graph, 1, 4);
    addEdge(graph, 2, 5);
    addEdge(graph, 2, 6);

    printf("Depth-First Search starting from vertex 0: ");
    DFS(graph, 0);
    printf("\n");

    return 0;
}
